guillotina/asgi.py

```diff
@@ -0,0 +1,118 @@
+from guillotina import glogging
+from guillotina import task_vars
+from guillotina.exceptions import ConflictError
+from guillotina.exceptions import TIDConflictError
+from guillotina.request import Request
+from guillotina.response import ASGISimpleResponse
+
+import asyncio
+
+
+logger = glogging.getLogger("guillotina")
+
+
+class AsgiApp:
+    def __init__(self, config_file, settings, loop):
+        self.app = None
+        self.config_file = config_file
+        self.settings = settings
+        self.loop = loop
+        self.on_cleanup = []
+        self.route = None
+        self.initialized = False
+
+    def __call__(self, scope, receive=None, send=None):
+        """
+        ASGI callable compatible with versions 2 and 3
+        """
+        if receive is None or send is None:
+
+            async def run_asgi2(receive, send):
+                return await self.real_asgi_app(scope, receive, send)
+
+            return run_asgi2
+        else:
+            return self.real_asgi_app(scope, receive, send)
+
+    async def real_asgi_app(self, scope, receive, send):
+        if scope["type"] == "http" or scope["type"] == "websocket":
+            return await self.handler(scope, receive, send)
+
+        elif scope["type"] == "lifespan":
+            while True:
+                message = await receive()
+                if message["type"] == "lifespan.startup":
+                    await self.startup()
+                    await send({"type": "lifespan.startup.complete"})
+                elif message["type"] == "lifespan.shutdown":
+                    await self.shutdown()
+                    await send({"type": "lifespan.shutdown.complete"})
+                    return
+
+    async def startup(self):
+        try:
+            from guillotina.factory.app import startup_app
+
+            self.loop = self.loop or asyncio.get_event_loop()
+
+            self.app = await startup_app(
+                config_file=self.config_file, settings=self.settings, loop=self.loop, server_app=self
+            )
+            self.initialized = True
+            return self.app
+        except Exception as e:
+            logger.exception("Something crashed during app startup")
+            raise e
+
+    async def shutdown(self):
+        for clean in self.on_cleanup:
+            await clean(self)
+
+    async def handler(self, scope, receive, send):
+        if not self.initialized:
```

Comment:
Maybe a lock/queue in case there is lot of traffic and waits that its initialized?

Reply:
It's not needed. An ASGI server that implements the lifespan events will take care to call the init before accepting requests and calling the handler. This conditional was added because the ASGI server `daphne` doesn't implement the lifespan events (https://github.com/django/daphne/issues/264) so we can fail with a proper error.

I'll add a comment to the code.

Reply:
done